lib/Sema/CSSimplify.cpp

```diff
@@ -6536,6 +6544,12 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
         // trivially solved.
         return getTypeMatchSuccess();
       }
+
+      // If exactly one of the type variables can bind to an pack, we
+      // can't merge these two type variables.
+      if (rep1->getImpl().canBindToPack()
+            != rep2->getImpl().canBindToPack())
+        return getTypeMatchFailure(locator);
```

Comment:
Instead of failing should this just from an unsolved constraint and wait until one of type variables is resolved so it could be fixed?

Reply:
TVO_CanBindToPack should never change. It's set when the type variable is created in CSGen.

Reply:
That's reasonable but we cannot diagnose here and at the same time cannot fail right away either I think, we need to waiting into more is know about one of the types or it's marked as a placeholder...

Reply:
Right, we can't unconditionally fail. When Pavel says "so it could be fixed", I think he means that this constraint needs to ultimately return `SolutionKind::Solved` in diagnostic mode with some kind of fix recorded. The solver should never fail solving a constraint in diagnostic mode.

Reply:
> should never fail solving a constraint in diagnostic mode

Shouldn't fail unless it leads to a worse solution i.e. if the impact of the fix is greater than previous solutions that have been discovered.